### PR TITLE
Export citation popup

### PR DESCRIPTION
**Pull request by mbollmann**
~This implements a feature discussed in #412 and supersedes #1095.~ (this now does none of these things)

This implements a new citation popup modal, making the interface nicer and giving us more options for hopefully providing more citation options soon (re #412).

Feedback welcome!

### CODE REVIEW

**Comment by mjpost** [comment]
This is fantastic. Comments:

* In the markdown text, there is a small gap between the pieces of the link: `[] ()`
* Copying currently requires three clicks: Cite → [tab] → Copy to clipboard. What do you think about retaining this (for discoverability), but also adding four tiny buttons beneath the "Cite" button (on the main paper page) that does each of them directly? That way, after the first time, users would just learn which to click. There are probably suitable icons for these four.
* Should the "Bib export formats" line below copy this functionality? It seems odd that we repeat just the "Copy BibTeX to clipboard" line

**Comment by mbollmann** [comment]
+ I think you're seeing the space before the (author, venue) part; there's no gap within the markdown link for me.
+ What about changing the "Bib export formats" line to have "copy to clipboard" options for each, as a quick access option for these? That way users would only have to open the "cite" dialog if they wanted the file download, or a preview of the texts before copying.
+ Also, there's no plain text / rich text version (as originally requested in #412 ...) yet. I'm not sure if this should be another tab in the dialog box or a line below the abstract, [as suggested here](https://github.com/acl-org/acl-anthology/pull/1095#issuecomment-733974853). What do you think?

**Comment by mjpost** [comment]
1. Oh, you're right!
2. I like it.
3. Could it be both or will that get too crowded?

Very excited about this change.

**Comment by mjpost** [comment]
A further thought: on (3), I like the idea of the inline "Short citation" text, maybe with its own "copy to clipboard" button. We only need to list it once, as rich text; the user can then use system options to paste as rich text or plain text ("paste and match style" on a mac).

**Comment by mjpost** [comment]
I'll add, that if the feedback complicates this PR, such that it gets stalled, let's just postpone them. This is a fantastic improvement and I'd love to see it get pushed out this week!

**Review by mjpost — APPROVED** [review]

**Comment by mbollmann** [comment]
Give me a day though to at least add some kind of plain text citation ;)

**Comment by mbollmann** [comment]
grafik

grafik

**Comment by akoehn** [comment]
Looks good, I just would switch markdown and bibtex, as bibtex is probably the dominant export format.

**Comment by mjpost** [comment]
Increasingly picky comments that can be ignored.

* The “Cite” button at the top stands out from “PDF” and “Search” in a negative way, to my eye. Not sure why it looks different
* “Citation (short)” feels wordy to me, and also redundant with the “Copy Citation” block above it. I’d suggest just calling it “Citation”, and moving it just above the “Copy Citation” block
* We could also also add “RTF” to the list of export options for completeness?

**Comment by mbollmann** [comment]
>     * The “Cite” button at the top stands out from “PDF” and “Search” in a negative way, to my eye. Not sure why it looks different

Is it just an aesthetic thing or does it actually look different for you? If the former, might be related to the width of the icon compared to the text. Maybe we could also use something else here now that the button doesn't actually download a file.

This here would be more in line with the citation symbol other websites (e.g. Google/Semantic Scholar) use:
button with quotation mark icon

Or this one:
button with pencil icon

Could of course also make the citation button primary-colored so that it stands out more:
button with primary color

>     * “Citation (short)” feels wordy to me, and also redundant with the “Copy Citation” block above it. I’d suggest just calling it “Citation”, and moving it just above the “Copy Citation” block

I'm thinking this is a very short and informal format, suitable for an email or blog post, but I wouldn't want to suggest that this is how we encourage papers in the Anthology to be cited. "short" would therefore be in contrast to the longer, ACL-style citation string that we can hopefully also add sometime soon.

>     * We could also also add “RTF” to the list of export options for completeness?

The button row, or the popup? Would it do the same thing the small copy button after "citation (short)" does right now?

**Comment by mjpost** [comment]
It looks quite different:

image

I like the cite with the double quotes! Probably we should color it as you displayed.

How about "Short citation" instead of "Citation (short)"?

I was thinking the popup and the button row, but I think we can leave it as you have it.

**Comment by nschneid** [comment]
I think only the PDF button should be blue. That is the main entry point to the content, so it should stand out.

**Comment by akoehn** [comment]
> It looks quite different

It does not look like that on any of my browsers; maybe it is a caching issue on your end @mjpost?

I like the " as well on the button.

"citation (short)" -> "short citation" -> "short cite" ?

Agree with @nschneid that the PDF should be the primary highlight.

**Comment by mbollmann** [comment]
@mjpost What's your browser/OS? Could you open the inspector (or whatever it's called in your browser) and modify that button's `<a ...>` element to

1. Include `href="#"`
2. Remove `type="button"`

and see if either of the two changes the display for you?

**Comment by mjpost** [comment]
Hmm, it’s all my devices (Macbook, iPad). Here’s on my phone (sorry can’t paste image from clipboard on iPad):

https://share.icloud.com/photos/0hpI-eBAhWK23B_Bcagk9R1hw

**Comment by mjpost** [comment]
I can do that mod but not till later tonight.

**Comment by mbollmann** [comment]
I'll just push a fix with (1) in a bit and you can check if that changes things. That actually seems to be a subtle styling bug that doesn't affect the display on my part unless I change the button's color, but maybe it causes a bigger difference on certain browsers.

**Comment by davidweichiang** [comment]
Sorry I'm late to this discussion, but I'm very worried that calling the "Title (Author, Conference Year)" format a citation of any kind will mislead people into thinking this is a standard citation format.

OTOH, I think people might appreciate a button that copies the BibTeX key to the clipboard.

**Comment by mbollmann** [comment]
> Sorry I'm late to this discussion, but I'm very worried that calling the "Title (Author, Conference Year)" format a citation of any kind will mislead people into thinking this is a standard citation format.

Would you feel that this would be less of an issue if this format was offered _alongside_ a proper ACL-style citation string?

> OTOH, I think people might appreciate a button that copies the BibTeX key to the clipboard.

Agreed, but we should tackle #1236 (which I'd like to do soon) before offering that.

**Comment by mjpost** [comment]
David, what effect of this do you fear? Does it help if we qualify it with "short" or "text" or "convenience" or something?

**Comment by davidweichiang** [comment]
I am afraid that people will write things like "In A Non-Linear Structural Probe (White et al., NAACL 2021), they do something similar to our paper, but also very different."

**Comment by davidweichiang** [comment]
> Does it help if we qualify it with "short" or "text" or "convenience" or something?

I am worried that it will be misunderstood no matter what qualifier is put on the front.

**Comment by mjpost** [comment]
I am having trouble imagining who would do this. Certainly not LaTeX users. So maybe Word users? I am not sure how References sections are generated in Word so maybe this could happen there. 

What if we called it a “blog citation”? Or “informal citation”? I think the utility here is high (for example, in writing short internal reports or notes, where I don’t generate a proper References section), and hope we can find a way to include it.

**Comment by aryamccarthy** [comment]
I agree with David about the risk. Is there any other anthology that provides this?

The benefit to providing BibTeX (or XML or whatever) is that it’s easy and common to mess up the format. If it’s meant to be informal for a blog, then there’s no format to mess up. People will just link to the article like they already do.

-am

**Comment by davidweichiang** [comment]
I don't know how one would check, but I would be completely unsurprised if a fraction of LaTeX users type the citations by hand. For sure, I've seen people write the .bbl files by hand.

**Comment by mjpost** [comment]
Semantic Scholar does these kinds of cites. We would call it the "Chicago" format. It'd be a good one to add (or MLA or APA). I've had separate requests privately for APA, now that this is mentioned.

image

**Comment by davidweichiang** [comment]
If it matches a standard citation format, then I think that's great.

**Comment by mbollmann** [comment]
We could use [citeproc-py](https://github.com/brechtm/citeproc-py/) to generate these different citation styles; you get all of the [styles from the CSL repository](https://github.com/citation-style-language/styles/) as options. I use this for the bibliography on my personal website, so I already know how it works.

One downside is that their "association-for-computational-linguistics" style appears to be broken for entries that contain a "month" field. So if we wanted to generate ACL-style citations with it, we'd need to update/maintain the CSL file for it as well. But maybe APA/Chicago/etc. is good enough too?

**Comment by mbollmann** [comment]
Okay, I just briefly tried to include citeproc-py locally, but it makes the build process take forever, so it doesn't seem to be a workable solution...

**Comment by mbollmann** [comment]
Nevermind, I just made a silly implementation mistake, sorry for the spam. :)

See [the latest commit](https://github.com/acl-org/acl-anthology/commit/fd2c9cd) for an example of what would be needed to integrate citeproc-py into the pipeline, to produce something like this (other citation styles of course possible as well):

grafik

**Comment by nschneid** [comment]
Nice, but if we're deeming it the ACL citation style, I think it should have the month as well.

**Comment by mjpost** [comment]
Also, 3+ authors should be converted to “et al.”, right?

**Comment by mbollmann** [comment]
> Also, 3+ authors should be converted to “et al.”, right?

No, that's the citation as it would appear under "References".

**Comment by mbollmann** [comment]
> Nice, but if we're deeming it the ACL citation style, I think it should have the month as well.

I guess we'd have to fix that [in the CSL file](https://github.com/citation-style-language/styles/blob/master/association-for-computational-linguistics.csl) :) ... currently it generates something like 

> ... pages 132–138, OnlineApril. Assocation for Computational Linguistics.

Or we "just" offer some variety of APA, MLA, Chicago. Although having our own citation style in there somewhere would be nice, I think.

**Comment by mjpost** [comment]
It is really nice having the ACL format. I see that in their ACL file, they note there is no official BST. That’s no longer correct, so we could set that straight.

**Comment by mbollmann** [comment]
> It is really nice having the ACL format. I see that in their ACL file, they note there is no official BST. That’s no longer correct, so we could set that straight.

Yes, looking at the commit history, it hasn't really been updated in 8 years. If we could have someone cross-check that CSL file with our latest BST and update it, that could be really useful.

**Comment by mjpost** [comment]
@nschneid It looks like the [ACL style file they use](https://github.com/citation-style-language/styles/blob/master/association-for-computational-linguistics.csl) is also used by Zotero. Do you know anything about this CSL format? Would you be willing to look into this?

**Comment by mjpost** [comment]
I'm not sure that the ACL style file actually does print the month. From a recent paper:

image

I am not familiar with the BST file format, but looking at [acl_natbib.bst](https://github.com/acl-org/ACLPUB/blob/master/templates/latex/acl_natbib.bst), it *seems* that the month only prints for web page and masters theses.

**Comment by nschneid** [comment]
CSL sounds familiar. I may have looked at this format but it was many years ago....anyway I would guess that the problem is

https://github.com/citation-style-language/styles/blob/master/association-for-computational-linguistics.csl#L264-L265

where the enclosing (`<if>`) block does not specify a delimiter to separate the 2 fields. I suspect wrapping them in `<group delimiter=", ">` will do the trick.

**Comment by mjpost** [comment]
One final thought: it also seems that the CSL fix might be easy: make [line 215](https://github.com/citation-style-language/styles/blob/52f9d99cc64c1642ed8dfae08ed4680a3114b138/association-for-computational-linguistics.csl#L215) look like [line 202](https://github.com/citation-style-language/styles/blob/52f9d99cc64c1642ed8dfae08ed4680a3114b138/association-for-computational-linguistics.csl#L202).

**Comment by nschneid** [comment]
> I am not familiar with the BST file format, but looking at [acl_natbib.bst](https://github.com/acl-org/ACLPUB/blob/master/templates/latex/acl_natbib.bst), it _seems_ that the month only prints for web pages and masters theses.

Interesting. It wouldn't hurt to print it whenever it's specified, right?

**Comment by mjpost** [comment]
Marcel, is there something you had to do to make citeproc fast? Instantiating the `Anthology` class now seems to take considerably more time. Perhaps lazy loading (#835) will be relevant here, too.

**Comment by mbollmann** [comment]
> Marcel, is there something you had to do to make citeproc fast? Instantiating the `Anthology` class now seems to take considerably more time. Perhaps lazy loading (#835) will be relevant here, too.

I made the mistake of re-loading the CSL file from disk (what's now [happening here](https://github.com/acl-org/acl-anthology/pull/1375/commits/fd2c9cd8aff5cfd5d3c3fa3fd29b354138130b40#diff-4f1686adf72e6df8f895dba7d5fbc812eecfe499839dbadee5174a78bf8773b6R51-R57)) for every single paper. I haven't benchmarked the exact overhead that adding citeproc-py now has, though.

What's still quite suboptimal is that those citations are all generated when a paper is instantiated; it would be better to only generate them when they're actually used. I believe that goes for other things in the codebase as well.

**Comment by mbollmann** [comment]
**tl;dr: Adding preformatted citations is not free of problems. Do you feel it's worth it to proceed with this?**

I've tweaked this a bit more so you can see what the code looks like & what this produces, currently only in the "Cite" popup, for example:

grafik

Some observations:

+ The overhead for instantiating the Anthology class was indeed significant, from 27 seconds (without the citeproc code) to 193 seconds for me. I've moved the citation string generation to `create_hugo_yaml.py` now, but it'll still have an effect on the full build time of course.
+ I can't find a CSL style that uses the short title fields for the proceedings (e.g. "EACL" instead of "Proceedings of the ..."). I think Semantic Scholar only does this because they don't even _have_ the full name of the proceedings in their bib info. If we want something like this, I think we'd have to construct a variant where we substitute the short conference title explicitly.
+ None of the other styles I've considered (Chicago, APA, MLA) appear to make use of the "month" field for conference paper entries.
+ The CSL styles are not free of problems:
    - ACL style has problems with "month" field (as reported before)
    - Chicago styles crash on page ranges with Roman numerals, so I left it out for now
    - MLA style with URL produces problematic HTML (unescaped angle brackets), so I used the version without the URL for now
    - Citeproc-py's HTML formatter does not add hyperlinks to the output, neither for titles nor for URLs

I'd love to hear your thoughts on whether it's worth it to proceed with this. I could also move the more controversial/diificult changes to a separate PR, so we could proceed with merging the new citation popup already.

**Comment by nschneid** [comment]
Seems to me like having multiple citation styles (MLA, Chicago, APA) is overkill—I don't expect the Anthology gets a ton of users who would use that (though who knows, I could be wrong).

The plain text ACL seems like a nice-to-have for putting together informal bibliographies (say, in a blog post or course document) that are not in LaTeX. A warning message (*__beta feature—may contain errors__*) could excuse having occasional errors in these.

**Comment by nschneid** [comment]
Short conference names are a separate issue (#567, #791). Would be great to have eventually, but I wouldn't hold up this feature over that.

**Comment by nschneid** [comment]
>  * ACL style has problems with "month" field (as reported before)

Have you tried the CSL fix I suggested above?

**Comment by akoehn** [comment]
I would propose we table the plaintext references for now and merge the easy (i.e. sone) parts.

Maybe we can find js bibtex renderer at some point in time and can mitigate the whole up-front generation.

**Comment by mjpost** [comment]
I like having an RTF citations, but agree we don't need all of them. I don't share David's concern about the text format, but would rather have consensus here before moving forward on this. Moreover, I'm working about increasing the instantiation time, since it really make debugging a pain. Perhaps we should fix #835 first.

So how about we factor the controversial stuff into another draft PR, and at least get the new popups live?

**Comment by mbollmann** [comment]
I agree that plain-text/rich-text ACL-style format would be the nicest. If that's all we need, maybe we can also build our own formatter for that in pure Python, to reduce overhead.

> Short conference names are a separate issue (#567, #791).

Yes, but tangential to this, as the CSL style files won't use them for anything.

> Have you tried the CSL fix I suggested above?

Not yet! We should, but only if we go forward with having citeproc-py in the pipeline.

> Moreover, I'm working about increasing the instantiation time, since it really make debugging a pain.

Instantiation time shouldn't be increased anymore with the latest commit here; overall build time is.

**Comment by mbollmann** [comment]
:arrow_right: Split off into #1390

**Comment by mjpost** [comment]
Oh, I thought we'd at least be keeping Markdown in this PR. Any objections to that? My fear is that #1390 is going to get stalled...

I'll lave the squash-merge to you, Marcel!

**Comment by mbollmann** [comment]
> Oh, I thought we'd at least be keeping Markdown in this PR. Any objections to that?

But isn't that exactly what David and others were concerned about?

**Comment by mjpost** [comment]
I thought David's concern stemmed from having a plain text "Short citation" format, inline on the main page. The Markdown format seems different to me (though maybe it's not in his mind).

**Comment by mbollmann** [comment]
> I thought David's concern stemmed from having a plain text "Short citation" format, inline on the main page. The Markdown format seems different to me (though maybe it's not in his mind).

It's exactly the same format, though? The "short citation" was just the rendered Markdown string. Of course it's less prominent if it only appears in the popup, and I could also include a footnote there that it's not an official citation format... I dunno, I'll go with whatever you (plural) feel is fine and not too problematic.

**Comment by mjpost** [comment]
The risk of cutting-and-pasting seems even lower to me (less visible, requires adjusting), but probably we should leave things as they are.

**Comment by davidweichiang** [comment]
Sorry, I hadn't looked at the Markdown before and would have the same opinion about the "short" citation in any format.

On my browser (Safari 14) the preview looks like this:

image

which is a little weird.

**Comment by mbollmann** [comment]
Maybe the Bootstrap version we use doesn't style `type="button"` correctly on Safari ... since I don't have an iOS device, could you check again now @davidweichiang?

**Comment by davidweichiang** [comment]
Yup, after a refresh that looks great.

**Comment by mjpost** [comment]
Same here.

**Comment by nschneid** [comment]
Beautiful! Announce the new feature on Twitter?

**Comment by davidweichiang** [comment]
Looks great.

Tangential question: What does the Search button do? Just searches on the title?

**Comment by mbollmann** [comment]
> Tangential question: What does the Search button do? Just searches on the title?

Yes exactly; on desktop the mouseover text will also say so. Opinions on whether or not to replace Google Scholar with Semantic Scholar via #1380, for the reasons expressed in #1339, are currently welcome!
